File: examples/03-intro-java/Example11/Example11.java
// =================================================================
//
// File: Example11.java
// Author(s):
//					 A01651517 Pedro Luis Gonzalez Roa 
//					 A01703947 Juan Alejandro Alcantara Minaya
//
// Description: This file implements the code that transforms a
//				grayscale image. The time this implementation takes will
//				be used as the basis to calculate the improvement obtained
// 				with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;

public class Example11 {
	private int src[], dest[], width, height;

	public Example11(int src[], int dest[], int width, int height) {
		this.src = src;
		this.dest = dest;
		this.width = width;
		this.height = height;
	}

	// place your code here

	void doTask() {
		// place your code here.
		for (int i = 0; i < src.length; i++) {
			int red = (src[i] & 0x00FF0000) >> 16; // AND mask & 2 byte shift
			int green = (src[i] & 0x0000FF00) >> 8; // AND mask & 1 byte shift
			int blue = (src[i] & 0x000000FF); // AND mask
			int avg = (red + green + blue)/3;
			dest[i] =
				0xFF000000 // ALPHA CHANNEL
				| (avg << 16) // RED CHANNEL
				| (avg << 8) // GREEN CHANNEL
				| avg // BLUE CHANNEL
				;
		}
	}

	public static void main(String args[]) throws Exception {
		long startTime, stopTime;
		double ms;

		if (args.length != 1) {
			System.out.println("usage: java Example11 image_file");
			System.exit(-1);
		}

		final String fileName = args[0];
		File srcFile = new File(fileName);
        final BufferedImage source = ImageIO.read(srcFile);

		int w = source.getWidth();
		int h = source.getHeight();
		int src[] = source.getRGB(0, 0, w, h, null, 0, w);
		int dest[] = new int[src.length];

		System.out.printf("Starting...\n");
		ms = 0;
		Example11 e = new Example11(src, dest, w, h);
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			e.doTask();

			stopTime = System.currentTimeMillis();

			ms += (stopTime - startTime);
		}
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
		final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
		destination.setRGB(0, 0, w, h, dest, 0, w);

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Original - " + fileName, source);
            }
        });

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Gray - " + fileName, destination);
            }
        });
	}
}

File: examples/02-intro-cpp/example11.cpp
// =================================================================
//
// File: example11.cpp
// Author(s):
// Description: This file implements the code that transforms a
//				grayscale image. Uses OpenCV, to compile:
//				g++ example11.cpp `pkg-config --cflags --libs opencv4`
//
//				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include "utils.h"


class GrayScale {
    private:
        cvv::Mat &src, &dest;

        void grayScalePixel(int ren, int col) const {
            int side_pixels, cells;
            int tmp_ren, tmp_col;
            float gray;

            dest.at<uchar>(ren, col) = (
                (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[RED];
                (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[GREEN];
                (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[BLUE];
            )/3;
        }

    public:
        GrayScale(cv::Mat &s, cv::Mat &d) : src(s), dest(d) {}

        void doTask() {
            for (int i = 0; i< src.rows; i++)
                for (int j = 0; j < src.cols; j++)
                    grayScalePixel(i, j);
        }
}

int main(int argc, char* argv[]) {
	int i;
	double acum;

	if (argc != 2) {
	printf("usage: %s source_file\n", argv[0]);
		return -1;
	}

	cv::Mat src = cv::imread(argv[1], cv::IMREAD_COLOR);
	cv::Mat dest = cv::Mat(src.rows, src.cols, CV_8UC3);
	if (!src.data) {
	printf("Could not load image file: %s\n", argv[1]);
		return -1;
	}

	acum = 0;
	for (i = 0; i < N; i++) {
		start_timer();

        GrayScale obj(src, dest);
        obj.doTask();

		acum += stop_timer();
	}

	printf("avg time = %.5lf ms\n", (acum / N));

	cv::namedWindow("Original", cv::WINDOW_AUTOSIZE);
	cv::imshow("Original", src);

	cv::namedWindow("Gray scale", cv::WINDOW_AUTOSIZE);
	cv::imshow("Gray scale", dest);

	cv::waitKey(0);

	cv::imwrite("gray_scale.png", dest);

	return 0;
}
